benchmark/src/main/java/feign/benchmark/FeignIteratorsBenchmark.java

```diff
@@ -0,0 +1,167 @@
+/**
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ * <p>
+ * http://www.apache.org/licenses/LICENSE-2.0
+ * <p>
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package feign.benchmark;
+
+import feign.Feign;
+import feign.RequestLine;
+import feign.jackson.JacksonDecoder;
+import feign.jackson.JacksonIteratorDecoder;
+import feign.stream.StreamDecoder;
+import io.netty.buffer.ByteBuf;
+import io.reactivex.netty.RxNetty;
+import io.reactivex.netty.protocol.http.server.HttpServer;
+import io.reactivex.netty.protocol.http.server.HttpServerRequest;
+import io.reactivex.netty.protocol.http.server.HttpServerResponse;
+import io.reactivex.netty.protocol.http.server.RequestHandler;
+import okhttp3.OkHttpClient;
+import org.openjdk.jmh.annotations.*;
+
+import java.util.Iterator;
+import java.util.List;
+import java.util.concurrent.Callable;
+import java.util.concurrent.TimeUnit;
+import java.util.stream.Stream;
+
+/**
+ * This test shows up memory consumption of different json array response processing implementations.
+ */
+@State(Scope.Thread)
+public class FeignIteratorsBenchmark {
+  private static final String HUNDRED = "100";
+  private static final String MILLION = "1000000";
+
+  interface ListApi {
+    @RequestLine("GET /cars/{count}")
+    List<Car> getCars(@feign.Param("count") String count);
+  }
+
+  interface IteratorApi {
+    @RequestLine("GET /cars/{count}")
+    Iterator<Car> getCars(@feign.Param("count") String count);
+  }
+
+  interface StreamApi {
+    @RequestLine("GET /cars/{count}")
+    Stream<Car> getCars(@feign.Param("count") String count);
+  }
+
+  @Param({"list", "iterator", "stream"})
+  private String api;
+
+  @Param({HUNDRED, MILLION})
+  private String size;
+
+  private Callable<Iterator<Car>> cars;
+
+  @Benchmark
+  @Warmup(iterations = 5, time = 1)
+  @Measurement(iterations = 10, time = 1)
+  @Fork(value = 3, jvmArgs = {"-server", "-Xms1G", "-Xmx1G"})
+  @BenchmarkMode(Mode.AverageTime)
+  @OutputTimeUnit(TimeUnit.NANOSECONDS)
+  public void feign() throws Exception {
+    final Iterator<Car> cars = this.cars.call();
+    while (cars.hasNext()) {
+      cars.next();
+    }
+  }
+
+  private static final int SERVER_PORT = 8765;
+  private HttpServer<ByteBuf, ByteBuf> server;
+
+  @Setup
+  public void setup() {
```

Comment:
actually I am going to remove this benchmark. You can keep it for your own purposes, but this is a heavy test just to look at decoder performance. In hind sight "real benchmarks" in JMH aren't great as they mislead. They are at best a hint (ex use wrk or something and you'll notice a huge difference on apps assembled and impact of things together.

If we were to make a iterator based thing, I'd weave it into the main real benchmark

Reply:
actually, please remove this on your own, as well the author tags. there are a fair amount of cleanups which I will explain in a summary comment.